Does not work Cursor using annotation based Mapper

<!--
Thank you for your contribution!

Question?
======================
Please use the mailing list. http://groups.google.com/group/mybatis-user
Questions on this tracker will be closed without comment.

Bug report?
======================
Please fill out the BUG REPORT FORM below.

To fix a bug, we need to reproduce it first.
And we spend a lot of time just trying to reproduce the reported problem, so please consider creating a failing test case or an example project.

- How to create a test case : https://github.com/mybatis/mybatis-3/wiki/Unit-Test
- How to create a good example : http://sscce.org
- How to upload your project to GitHub: https://help.github.com/articles/adding-an-existing-project-to-github-using-the-command-line/

Feature request?
=======================
- Please delete the BUG REPORT FORM below and describe the feature.
- It is a good idea to discuss your changes on the mailing list to get feedback from the community. http://groups.google.com/group/mybatis-user
- If you have a patch with unit tests, send a pull request. Please see the wiki page : https://github.com/mybatis/mybatis-3/wiki/Contribute
-->

<!-- BUG REPORT FORM -->

## MyBatis version
3.4.0

## Database vendor and version
Any

## Test case or example project

N/A

## Steps to reproduce

At first, i add method in `BoundBlogMapper` as follow:

```java
  @Select({
          "SELECT *",
          "FROM blog",
          "ORDER BY id"
  })
  Cursor<Blog> selectCursorBlogs();
```

At second, i add test method in `BindingTest` as follow:

```java
  @Test
  public void shouldExecuteWithCursorAndRowBounds() {
    SqlSession session = sqlSessionFactory.openSession();
    try {
      BoundBlogMapper mapper = session.getMapper(BoundBlogMapper.class);
      Cursor<Blog> blogs = mapper.selectCursorBlogs();
      for (Blog b :blogs) {
        System.out.println(b);
      }
    } finally {
      session.close();
    }
  }
```

## Expected result

```
Blog: 1 : Jim Business (null)
Blog: 2 : Bally Slog (null)
```

## Actual result

```
org.apache.ibatis.reflection.ReflectionException: Error instantiating interface org.apache.ibatis.cursor.Cursor with invalid types () or values (). Cause: java.lang.NoSuchMethodException: org.apache.ibatis.cursor.Cursor.<init>()

	at org.apache.ibatis.reflection.factory.DefaultObjectFactory.instantiateClass(DefaultObjectFactory.java:90)
	at org.apache.ibatis.reflection.factory.DefaultObjectFactory.create(DefaultObjectFactory.java:50)
	at org.apache.ibatis.reflection.factory.DefaultObjectFactory.create(DefaultObjectFactory.java:42)
	at org.apache.ibatis.executor.resultset.DefaultResultSetHandler.createResultObject(DefaultResultSetHandler.java:594)
	at org.apache.ibatis.executor.resultset.DefaultResultSetHandler.createResultObject(DefaultResultSetHandler.java:571)
	at org.apache.ibatis.executor.resultset.DefaultResultSetHandler.getRowValue(DefaultResultSetHandler.java:380)
	at org.apache.ibatis.executor.resultset.DefaultResultSetHandler.handleRowValuesForSimpleResultMap(DefaultResultSetHandler.java:339)
	at org.apache.ibatis.executor.resultset.DefaultResultSetHandler.handleRowValues(DefaultResultSetHandler.java:314)
	at org.apache.ibatis.cursor.defaults.DefaultCursor.fetchNextObjectFromDatabase(DefaultCursor.java:142)
	at org.apache.ibatis.cursor.defaults.DefaultCursor.fetchNextUsingRowBound(DefaultCursor.java:128)
	at org.apache.ibatis.cursor.defaults.DefaultCursor$CursorIterator.hasNext(DefaultCursor.java:199)
	at org.apache.ibatis.binding.BindingTest.shouldExecuteWithCursorAndRowBounds(BindingTest.java:774)
...
Caused by: java.lang.NoSuchMethodException: org.apache.ibatis.cursor.Cursor.<init>()
	at java.lang.Class.getConstructor0(Class.java:3082)
	at java.lang.Class.getDeclaredConstructor(Class.java:2178)
	at org.apache.ibatis.reflection.factory.DefaultObjectFactory.instantiateClass(DefaultObjectFactory.java:62)
```

--- a/src/main/java/org/apache/ibatis/builder/annotation/MapperAnnotationBuilder.java
+++ b/src/main/java/org/apache/ibatis/builder/annotation/MapperAnnotationBuilder.java
@@ -62,6 +62,7 @@ import org.apache.ibatis.builder.BuilderException;
 import org.apache.ibatis.builder.IncompleteElementException;
 import org.apache.ibatis.builder.MapperBuilderAssistant;
 import org.apache.ibatis.builder.xml.XMLMapperBuilder;
+import org.apache.ibatis.cursor.Cursor;
 import org.apache.ibatis.executor.keygen.Jdbc3KeyGenerator;
 import org.apache.ibatis.executor.keygen.KeyGenerator;
 import org.apache.ibatis.executor.keygen.NoKeyGenerator;
@@ -390,7 +391,7 @@ public class MapperAnnotationBuilder {
     } else if (resolvedReturnType instanceof ParameterizedType) {
       ParameterizedType parameterizedType = (ParameterizedType) resolvedReturnType;
       Class<?> rawType = (Class<?>) parameterizedType.getRawType();
-      if (Collection.class.isAssignableFrom(rawType)) {
+      if (Collection.class.isAssignableFrom(rawType) || Cursor.class.isAssignableFrom(rawType)) {
         Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
         if (actualTypeArguments != null && actualTypeArguments.length == 1) {
           Type returnTypeParameter = actualTypeArguments[0];

--- a/src/test/java/org/apache/ibatis/submitted/cursor_simple/AnnotationMapper.java
+++ b/src/test/java/org/apache/ibatis/submitted/cursor_simple/AnnotationMapper.java
@@ -1,0 +1,26 @@
+/**
+ *    Copyright 2009-2016 the original author or authors.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package org.apache.ibatis.submitted.cursor_simple;
+
+import org.apache.ibatis.annotations.Select;
+import org.apache.ibatis.cursor.Cursor;
+
+public interface AnnotationMapper {
+
+  @Select("select * from users order by id")
+  Cursor<User> getAllUsers();
+
+}

--- a/src/test/java/org/apache/ibatis/submitted/cursor_simple/CursorSimpleTest.java
+++ b/src/test/java/org/apache/ibatis/submitted/cursor_simple/CursorSimpleTest.java
@@ -1,5 +1,5 @@
 /**
- *    Copyright 2009-2015 the original author or authors.
+ *    Copyright 2009-2016 the original author or authors.
  *
  *    Licensed under the Apache License, Version 2.0 (the "License");
  *    you may not use this file except in compliance with the License.
@@ -30,7 +30,9 @@ import org.junit.Test;
 import java.io.IOException;
 import java.io.Reader;
 import java.sql.Connection;
+import java.util.ArrayList;
 import java.util.Iterator;
+import java.util.List;
 
 public class CursorSimpleTest {
 
@@ -294,4 +296,41 @@ public class CursorSimpleTest {
 
         Assert.fail("Should have returned earlier");
     }
+
+    @Test
+    public void shouldGetAllUserUsingAnnotationBasedMapper() {
+        SqlSession sqlSession = sqlSessionFactory.openSession();
+        sqlSession.getConfiguration().getMapperRegistry().addMapper(AnnotationMapper.class);
+        AnnotationMapper mapper = sqlSession.getMapper(AnnotationMapper.class);
+        Cursor<User> usersCursor = mapper.getAllUsers();
+        try {
+            Assert.assertFalse(usersCursor.isOpen());
+            Assert.assertFalse(usersCursor.isConsumed());
+            Assert.assertEquals(-1, usersCursor.getCurrentIndex());
+
+            List<User> userList = new ArrayList<User>();
+            for (User user : usersCursor){
+                userList.add(user);
+                Assert.assertEquals(userList.size() -1, usersCursor.getCurrentIndex());
+            }
+
+            Assert.assertFalse(usersCursor.isOpen());
+            Assert.assertTrue(usersCursor.isConsumed());
+            Assert.assertEquals(3, usersCursor.getCurrentIndex());
+
+            Assert.assertEquals(4, userList.size());
+            User user = userList.get(0);
+            Assert.assertEquals("User1", user.getName());
+            user = userList.get(1);
+            Assert.assertEquals("User2", user.getName());
+            user = userList.get(2);
+            Assert.assertEquals("User3", user.getName());
+            user = userList.get(3);;
+            Assert.assertEquals("User4", user.getName());
+
+        } finally {
+            sqlSession.close();
+        }
+    }
+
 }

